### ISSUE TITLE
Vcat automatic recognition of same columns name

### ISSUE THREAD

**Issue by DarioSarra** [JuliaData/DataFrames.jl#1347]
I tried to look a lot for this method and it feels like in a previous version this was working, I hope I am not missing the existence of it. 
However:
I am trying to use vcat to combine different DataFrames, which have the same columns but in different orders. Wouldn't be nice if vcat could access the name list and mutually combine the columns based on the names instead of looking ony for the same exact order?
minimal example

```julia
a=DataFrame(x = [1,2,3], y = ["a","b","c"])
b=DataFrame( y = ["d","e","f"], x = [4,5,6])
vcat(a,b)
```

**Comment by nalimilan** [comment]
That kind of makes sense (and that's how R and dplyr work with `rbind`). Though there's a tension between treating data frames as a vector of columns with an order or as a matrix, and treating them like a database. `vcat` is essentially a vector/matrix API, so we could also provide a different function to combine columns by name.

AFAICT the error was introduced by https://github.com/JuliaData/DataTables.jl/pull/45, but it doesn't seem we've discussed the opportunity of changing the behavior. Cc: @cjprybol

**Comment by tbeason** [comment]
I also ran into this recently. I think it makes sense to provide a fallback for if the `df` columns are in a different order, perhaps using a permutation on `names`? Might be slightly slower, but could just document that it will be much faster if the columns are similarly ordered on the two (or many) dataframes.

For example, ignoring the necessary error-checking, something like the below would do the trick. 
```julia
permcols(dfa,dfb) = collect( findfirst(equalto(nma),names(dfb)) for nma in names(dfa))
``` 
Then `dfb[permcols(dfa,dfb)]` results in `dfb` having the same column order as `dfa`

**Comment by nalimilan** [comment]
It's really not hard to do and performance should be the same, it's more of a design question.

**Comment by DarioSarra** [comment]
I solved the problem reorganising the second dataframe in the order of the first one with a simple function

`concat_data!(a, b) = append!(a, b[:, names(a)])`

I guess this could be add as an alternative method if the function checks:
1 mismatch in col names
2 presence of the same names

**Comment by nalimilan** [comment]
@bkamins Would you be willing to work on this?

**Comment by bkamins** [comment]
I can do it, but first let us agree on the contract we want to follow comprehensively for those functions. Then we can decide on the implementation plan.

We have two functions that combine `DataFrames`:
* `vcat` creates a new `DataFrame`, can join more than two `DataFrames`, promotes types of columns;
* `append!` appends to the first `DataFrame`, accepts only two arguments, requires strictly the same types of both columns (this is different to the behavior of `append!` in base which performs the conversion if possible).

Here is the list of things I think that can be done. Can you please indicate which ones would be suitable:
1. add `reorder` keyword argument to `vcat` with default to `false`; if it is set to `true` we try to match columns; in this way we will not break `[df1; df2]` syntax which would still work in matrix-oriented way; (maybe a better name can be proposed)
2. add `reorder` keyword argument to `append!` with the same logic;
3. allow more than one `DataFrame` to be appended in `append!` (like in `vcat`);
4. allow promotion of appended columns in `append!`.

Points 3 and 4 are potentially important because broadcasting `append!.(df1, [df2, df3])` is not atomic.

**Comment by nalimilan** [comment]
Cool!

> add reorder keyword argument to vcat with default to false; if it is set to true we try to match columns; in this way we will not break [df1; df2] syntax which would still work in matrix-oriented way; (maybe a better name can be proposed)
>     add reorder keyword argument to append! with the same logic;

I'm not sure what's the "matrix-oriented way". Do you mean that columns should be matched by position even if their names differ? That doesn't sound too useful and I don't think other data frame software allow this. I'd rather always use the names, applying the order of the first data frame (I know my last comment wasn't clearly in favor of this but...).

>     allow more than one DataFrame to be appended in append! (like in vcat);

I see your point about atomicity. I'm tempted to defer this issue as it's not completely related, and because there's no equivalent in the generic Base method. We could discuss adding it both to Base and to DataFrames.

>     allow promotion of appended columns in append!.

+1. But that's not strictly required to fix `vcat` either.

**Comment by bkamins** [comment]
We can defer the other two issues. Regarding `reorder` keyword argument:

1. Do we agree to add it to `vcat` and `append!` or only `vcat`?
2. Do we agree that `reorder` is the name to use
2. What I meant is:
    * if `reorder=true` then we check if all `DataFrame`s have the same names of columns but they do not have to be in the same order; if this is the case we allow reordering the names of all `DataFrame`s to match the first `DataFrame`;
    * if `reorder=false` (the default) then we keep current functionality: we check if all `DataFrames` have the same names of columns and in the same order (that is what I meant by matrix-oriented: that everything must match).

**Comment by nalimilan** [comment]
I think we should go without the keyword argument and reorder columns by default. That sounds simpler, I'm not sure there's a real interest in being so strict.

**Comment by DarioSarra** [comment]
I would suggest to add this functionality to append!, the reason stems from the difference between vcat and append! pointed by @bkamins. If the user calls vcat probably wants to create a new variable if calls append! the 2 element in his mind should be portion of a single larger dataframe. About the keyword argument I think it would be good to being able to choose, but the default could be  `reorder = true`

**Comment by bkamins** [comment]
I have implemented it as @nalimilan suggested in PR #1366 (only `vcat` and without keyword argument). Of course it can be modified if we decide we want to add keyword argument (note that in `[df1; df2]` syntax we are not able to pass this keyword argument to `vcat`).

**Comment by nalimilan** [comment]
Fixed by https://github.com/JuliaData/DataFrames.jl/pull/1366.
